Added an option for multiple initial conditions in IDAKLU solver

Rishab87 opened the pull request:
# Description

Added an option for multiple initial conditions in IDAKLU solver

Fixes #3713 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #)

# Important checks:

Please confirm the following before marking the PR as ready for review:
- [x] No style issues: `nox -s pre-commit`
- [x] All tests pass: `nox -s tests`
- [x] The documentation builds: `nox -s doctests`
- [x] Code is commented for hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

kratman commented on CHANGELOG.md:
```diff
@@ -21,6 +21,8 @@
 - Added support for particle size distributions combined with particle mechanics. ([#4807](https://github.com/pybamm-team/PyBaMM/pull/4807))
 - Added InputParameter support in PyBamm experiments ([#4826](https://github.com/pybamm-team/PyBaMM/pull/4826))
 - Added support for the `"pchip"` interpolator using the CasADI backend. ([#4871](https://github.com/pybamm-team/PyBaMM/pull/4871))
+- Added "use lumped thermal capacity" option in lumped thermal model ([#4968](https://github.com/pybamm-team/PyBaMM/pull/4968))
+- Added an option for multiple initial conditions in IDAKLU solver ([#4981](https://github.com/pybamm-team/PyBaMM/pull/4981))
```
```suggestion
- Added "use lumped thermal capacity" option in lumped thermal model ([#4968](https://github.com/pybamm-team/PyBaMM/pull/4968))
- Added an option for multiple initial conditions in IDAKLU solver ([#4981](https://github.com/pybamm-team/PyBaMM/pull/4981))
```
Wrong section, this needs to go into unreleased.

kratman replied:
I also fixed the lumped thermal capacity option in the changelog for the last bugfix release, so just move the new one up to unreleased

Rishab87 replied:
Thanks for the review moved them

martinjrobins commented on src/pybamm/solvers/idaklu_solver.py:
```diff
@@ -861,11 +898,35 @@ def _integrate(self, model, t_eval, inputs_list=None, t_interp=None):
+                y0_list = []
+
+                for ic in initial_conditions:
+                    model_copy = model.new_copy()
```
you don't want to copy the model for each initial_condition, as models are quite big

martinjrobins commented on src/pybamm/solvers/base_solver.py:
```diff
@@ -739,7 +740,11 @@ def solve(
             The times (in seconds) at which to interpolate the solution. Defaults to None.
             Only valid for solvers that support intra-solve interpolation (`IDAKLUSolver`).
-
+        initial_conditions : dict, numpy.ndarray, or list, optional
```
put somewhere here that this is only for the idaklu solver

Rishab87 replied:
yeah right,added that

martinjrobins commented on src/pybamm/solvers/base_solver.py:
```diff
@@ -911,6 +916,7 @@ def solve(
                     t_eval[start_index:end_index],
                     model_inputs_list,
                     t_interp,
+                    initial_conditions,
```
does this arg work for the jax solver?

Rishab87 replied:
Not now , I'll add support for jax in a seperate PR in future otherwise this would become too long for now

martinjrobins commented on src/pybamm/solvers/idaklu_solver.py:
```diff
@@ -822,7 +822,44 @@ def supports_parallel_solve(self):
         return False
 
-    def _integrate(self, model, t_eval, inputs_list=None, t_interp=None):
+    def _apply_initial_conditions(self, model, initial_conditions):
```
this function overlaps a bit with `BaseModel.set_initial_conditions_from`, have you seen this function?

Rishab87 replied:
yeah I saw it but it works only when we have a solution object. In multiple IC scenario, we need to spin up copies of the discretised and build model before we call the IDA integrator so I think we should have a separate `_apply_initial_conditions` function?

martinjrobins replied:
happy to keep them separate, just wanted to point it out as this function does a lot of special handling for concatenated symbols which you might need for your `_apply_initial_conditions` (or might not?)

Rishab87 replied:
by the time we get to _apply_solver_initial_conditions, the model has already been discretized and flattened into a single y0 vector, so I dont think we need special handling for concatenated symbols

martinjrobins commented on tests/unit/test_solvers/test_idaklu_solver.py:
```diff
@@ -1298,3 +1297,211 @@ def test_model_solver_with_non_identity_mass(self):
+        n_sims = 3
+        initial_conditions = [{"u": i + 1} for i in range(n_sims)]
+        inputs = [{} for _ in range(n_sims)]
+        t_eval = np.linspace(0, 1, 10)
```
use:
```python
t_eval = [0, 1]
t_interp = np.linspace(0, 1, 10)
```

then

```python
solutions = solver.solve(
            model, t_eval, t_interp=t_interp, inputs=inputs, initial_conditions=initial_conditions
        )
```

kratman replied:
This change did not seem to be done yet

martinjrobins commented on tests/unit/test_solvers/test_idaklu_solver.py:
```diff
@@ -1298,3 +1297,211 @@ def test_model_solver_with_non_identity_mass(self):
+        disc = pybamm.Discretisation()
+        disc.process_model(model)
+
+        solver = pybamm.IDAKLUSolver(rtol=1e-8, atol=1e-10, options={"num_threads": 1})
```
I don't think you need these tolerances for such a simple problem, just use the defaults

Rishab87 replied:
without these getting assertion errors for these problems, alternatively we can add these tolerances while comparing if not here?

martinjrobins replied:
when comparing the solution? I would just use the default tolerances, then use these also (perhaps multipled by 10 or so) for your comparison. I notice you don't specify an absolute tolerance for your comparison, the default is 0 which is not suitable for a solution tending to 0.

martinjrobins commented on src/pybamm/solvers/jax_solver.py:
```diff
@@ -189,7 +189,9 @@ def supports_parallel_solve(self):
-    def _integrate(self, model, t_eval, inputs=None, t_interp=None):
+    def _integrate(
+        self, model, t_eval, inputs=None, t_interp=None, intial_conditions=None
+    ):
```
If you are not planning on supporting the jax solver, you need to raise an error if `initial_conditions` is not None

martinjrobins reviewed (changes requested):
thanks @Rishab87, this is a great start and on the right track. I've made a few comments and suggestions below, happy to discuss any of them just reply to this or message me on the slack.

Rishab87 commented on tests/integration/test_solvers/test_idaklu.py:
```diff
@@ -208,3 +208,39 @@ def test_with_experiments(self):
             sols[1].cycles[-1]["Current [A]"].data,
         )
+
+    def test_multiple_initial_conditions(self):
```
added them

Rishab87 commented:
@martinjrobins Thanks for the review, I've made few changes accordingly and replied to all the comments

martinjrobins reviewed (changes requested):
thanks @Rishab87, `test_multiple_initial_conditions_spm` and `test_multiple_initial_conditions_dfn` look very similar, can we combine this into a single test that is parameterised by the model class? Also, it would be good to check the solution against two separate solves using those initial conditions, so we know the solution is correct, not just the initial condition

kratman commented on CHANGELOG.md:
```diff
@@ -25,6 +25,8 @@
 ## Features
 
 - Added "use lumped thermal capacity" option in lumped thermal model ([#4968](https://github.com/pybamm-team/PyBaMM/pull/4968))
+- Added an option for multiple initial conditions in IDAKLU solver ([#4981](https://github.com/pybamm-team/PyBaMM/pull/4981))
```
Move this to unreleased above

Rishab87 commented:
@martinjrobins thanks for the review again, I've made few changes in the tests and replied to the comments

martinjrobins reviewed (approved):
thanks @Rishab87 for making the changes, sorry for the delay in review this fell off my radar. All looks good to me now

kratman commented on CHANGELOG.md:
```diff
@@ -32,6 +32,10 @@
 - Fixed a bug in the `Exponential1DSubMesh` where the mesh was not being created correctly for non-zero minimum values. ([#4989](https://github.com/pybamm-team/PyBaMM/pull/4989))
 - Fixed sensitivity calculation for `pybamm.DiscreteTimeSum`. ([#5007](https://github.com/pybamm-team/PyBaMM/pull/5008))
 
+## Features
+
+- Added an option for multiple initial conditions in IDAKLU solver ([#4981](https://github.com/pybamm-team/PyBaMM/pull/4981))
+
```
This needs to be moved up to the correct release now. Pains of working with a manual change log

kratman commented on src/pybamm/solvers/idaklu_solver.py:
```diff
@@ -838,6 +875,11 @@ def _integrate(self, model, t_eval, inputs_list=None, t_interp=None):
+            - a dict mapping variable names → values
+            - a 1D array of length `n_states`
+            - a list of such overrides (one per parallel solve)
         """
```
These blocks do not render correctly in sphinx. Usually the fix is to have black lines before and/or after the list. You can view the generated docs in either the RTD CI or by running locally when you fix this.

![image](https://github.com/user-attachments/assets/d5cb0688-ed7d-418c-9b09-802e23375b06)

kratman replied:
Note: I mentioned it on this one, but I have seen this block several times

Rishab87 replied:
This renders correctly now:

<img width="750" height="176" alt="image" src="https://github.com/user-attachments/assets/fdafe4d3-366c-4b3f-8ab3-a0128e122ad6" />

Rishab87 commented:
@kratman any updates on this?

kratman commented on tests/unit/test_solvers/test_idaklu_solver.py:
```diff
@@ -1298,3 +1297,211 @@ def test_model_solver_with_non_identity_mass(self):
+        ):
+            solver.solve(
+                model, t_eval, inputs=inputs, initial_conditions=initial_conditions
+            )
```
@martinjrobins Do we need a ticket for this or is this resolved

kratman commented on src/pybamm/solvers/idaklu_solver.py:
```diff
@@ -565,11 +609,36 @@ def _integrate(self, model, t_eval, inputs_list=None, t_interp=None):
+        else:
+            # stack y0full and ydot0full so they are a 2D array of shape (number_of_inputs, number_of_states + number_of_parameters * number_of_states)
+            # note that y0full and ydot0full are currently 1D arrays (i.e. independent of inputs), but in the future we will support
+            # different initial conditions for different inputs (see https://github.com/pybamm-team/PyBaMM/pull/4260). For now we just repeat the same initial conditions for each input
```
I know this was not your change, it just got moved without changing something that should have changed

Rishab87 commented:
@kratman I've already addressed open items here:
- used `t_eval` & `t_interp` as told there in all the unit tests.
- used reasonable tolerances in unit tests `rtol=1e-3` & `atol=1e-5`.  
- Added integration tests for PDE models (SPM & DFN).

Anything else you would like to know?

kratman commented:
> @kratman I've already addressed open items here:
> 
> * used `t_eval` & `t_interp` as told there in all the unit tests.
> * used reasonable tolerances in unit tests `rtol=1e-3` & `atol=1e-5`.
> * Added integration tests for PDE models (SPM & DFN).
> 
> Anything else you would like to know?

I think a ticket for the Jax solver portion would be good

kratman commented:
> @kratman yeah sure I'll keep jax seperate this PR does not have any jax related code, by ticket you mean should I create a seperate issue for jax?

Yes
